Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp

```diff
@@ -33,43 +40,146 @@ static std::string GetJoystickName(int index)
 #endif
 }
 
-void Init()
+static void OpenAndAddDevice(int index)
 {
-#ifdef USE_SDL_HAPTIC
-  if (SDL_Init(SDL_INIT_JOYSTICK | SDL_INIT_HAPTIC) >= 0)
+  SDL_Joystick* dev = SDL_JoystickOpen(index);
+  if (dev)
   {
-    // Correctly initialized
+    auto js = std::make_shared<Joystick>(dev, index);
+    // only add if it has some inputs/outputs
+    if (!(js->Inputs().empty() && js->Outputs().size()))
+      g_controller_interface.AddDevice(std::move(js));
   }
-  else
-#endif
-      if (SDL_Init(SDL_INIT_JOYSTICK) < 0)
+}
+
+#if SDL_VERSION_ATLEAST(2, 0, 0)
+static Common::Event s_init_event;
+static Uint32 s_stop_event_type;
+static Uint32 s_populate_event_type;
+static Common::Event s_populated_event;
+static std::thread s_hotplug_thread;
+
+static bool HandleEventAndContinue(const SDL_Event& e)
+{
+  if (e.type == SDL_JOYDEVICEADDED)
   {
-    // Failed to initialize
-    return;
+    OpenAndAddDevice(e.jdevice.which);
+    g_controller_interface.InvokeHotplugCallbacks();
   }
+  else if (e.type == SDL_JOYDEVICEREMOVED)
+  {
+    g_controller_interface.RemoveDevice([&e](const auto* device) {
+      const Joystick* joystick = dynamic_cast<const Joystick*>(device);
+      return joystick && SDL_JoystickInstanceID(joystick->GetSDLJoystick()) == e.jdevice.which;
+    });
+    g_controller_interface.InvokeHotplugCallbacks();
+  }
+  else if (e.type == s_populate_event_type)
+  {
+    for (int i = 0; i < SDL_NumJoysticks(); ++i)
+      OpenAndAddDevice(i);
+    s_populated_event.Set();
+  }
+  else if (e.type == s_stop_event_type)
+  {
+    return false;
+  }
+
+  return true;
+}
+#endif
+
+void Init()
+{
+#if !SDL_VERSION_ATLEAST(2, 0, 0)
+  if (SDL_Init(SDL_INIT_JOYSTICK) != 0)
+    ERROR_LOG(SERIALINTERFACE, "SDL failed to initialize");
+  return;
```

Comment:
The return isn't needed here

Reply:
You're right, but I think it makes it a lot clearer, given the giant else block that comes after it.